File: app/src/main/java/de/threateningcodecomments/routinetimer/Routine.java
package de.threateningcodecomments.routinetimer;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.UUID;

class Routine {
    public static final Routine ERROR_ROUTINE = new Routine(Routine.ERROR_NAME,
            new ArrayList<>(
                    Arrays.asList(Tile.ERROR_TILE, Tile.ERROR_TILE)
            )
    );
    public static final String ERROR_NAME = Tile.ERROR_NAME;

    public static final int MODE_CONTINUOUS = 1;
    public static final int MODE_SEQUENTIAL = 0;

    private int mode;

    private String name;
    private ArrayList<Tile> tiles;

    private String UID;

    //debug constructor
    public Routine(int mode, String name, ArrayList<Tile> tiles) {
        this.mode = mode;
        this.name = name;
        this.tiles = tiles;

        if (this.UID == null) {
            this.UID = UUID.randomUUID().toString();
        }
    }

    public Routine(String name, ArrayList<Tile> tiles) {
        this.name = name;
        this.tiles = tiles;
    }

    public Routine() {
    }

    //region Getters and Setters
    public int getMode() {
        return mode;
    }

    public void setMode(int mode) {
        this.mode = mode;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public ArrayList<Tile> getTiles() {
        return tiles;
    }

    public void setTiles(ArrayList<Tile> tiles) {
        this.tiles = tiles;
    }

    public String getUID() {
        return UID;
    }

    public void setUID(String UID) {
        if (this.UID == null) {
            this.UID = UID;
        }
    }
    //endregion
}

File: app/src/main/java/de/threateningcodecomments/routinetimer/Routine.kt
package de.threateningcodecomments.routinetimer

import java.util.*
import kotlin.collections.ArrayList

class Routine {
    var name: String? = null
    var uid: String? = null
        get() = if (field == null) {
            val randomUid = UUID.randomUUID().toString()
            field = randomUid
            randomUid
        } else {
            field
        }
        set(value) {
            if (field == null) {
                field = value
            }
        }

    var lastUsed: Long? = null
        get() = if (field == null) {
            val currentTime = System.currentTimeMillis()
            field = currentTime
            field
        } else {
            field
        }

    var mode = 0
    var tiles: ArrayList<Tile>? = null

    //region Constructors
    constructor(mode: Int, name: String, tiles: ArrayList<Tile>, lastUsed: Long) {
        this.mode = mode
        this.name = name
        this.tiles = tiles
        if (uid == null) {
            uid = UUID.randomUUID().toString()
        }
        this.lastUsed = lastUsed
    }

    constructor(name: String?, uid: String, tiles: ArrayList<Tile>?) {
        this.name = name
        this.uid = uid
        this.tiles = tiles
    }

    constructor() {}

    constructor(name: String, uid: String?, mode: Int, tiles: ArrayList<Tile>) {
        this.name = name
        this.uid = uid
        this.mode = mode
        this.tiles = tiles
    }
    //endregion

    fun setAccessibility(isNightMode: Boolean) {
        if (tiles == null) {
            tiles = ArrayList()
        }

        for (tmpTile in tiles!!) {
            tmpTile.setAccessibility(isNightMode)
        }
    }

    companion object {
        const val ERROR_NAME = Tile.ERROR_NAME
        const val ERROR_UID = "69420"

        const val MODE_CONTINUOUS = 1
        const val MODE_SEQUENTIAL = 0

        const val CONTINUOUS_MESSAGE: String = "Continuous mode"
        const val SEQUENTIAL_MESSAGE: String = "Sequential mode"

        @JvmField
        val ERROR_ROUTINE = Routine(ERROR_NAME, ERROR_UID,
                ArrayList(
                        listOf(Tile.ERROR_TILE)
                )
        )
    }
}